---
title: "scMethrix Summary"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    self_contained: yes
    css: custom.css
    theme: sandstone
    highlight: tango
params:
  hasCov: FALSE
  plotDensityScore: NULL
  plotDensityCoverage: NULL
  plotCpGsCoveredCountBySample: NULL
  plotCpGsCoveredCountByChr: NULL
  plotCpGsCoveredFracBySample: NULL
  plotCpGsCoveredFracByChr: NULL
  plotCpGsCoveredCountBySampleAll: NULL
  plotCpGsCoveredCountByChrAll: NULL
  plotCpGsCoveredFracBySampleAll: NULL
  plotCpGsCoveredFracByChrAll: NULL
  plotMethylationAll: NULL,
  plotMethylation: NULL,
  plotCoverageAll: NULL,
  plotCoverage: NULL
---

```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
```

```{r echo=FALSE}
# Input a named list of plots, and this creates a tabbed panel of the plots in plotly format
putInTabs <- function(plots, level = 2) {
  
  level <- paste(rep("#",level), collapse="")
  
  for(plot in names(plots)) {
    cat(level, plot, '{-}',  '\n\n')
    print(htmltools::tagList(plotly::ggplotly(plots[[plot]])))
    cat( '\n\n')
  }
}
```

```{r, include=FALSE}
# To make sure plotly libs are initialized
htmltools::tagList(plotly::plot_ly(data = iris, x = ~Sepal.Length, y = ~Petal.Length))
```


<img src="logo_large_hexagon.gif" height="150" width="150" style="position:absolute;top:30px;right:0px;"/>

```{r}
if (!requireNamespace("ggplot2", quietly = TRUE))
    install.packages("ggplot2")
library(ggplot2)

if (!requireNamespace("plotly", quietly = TRUE))
    install.packages("plotly")
library(plotly)
```


<!-- Figure 1 -->

```{r, echo=FALSE}
plots <- list()
plots[["Score"]] <- params$plotDensityScore

if (params$hasCov) {
  plots[["Coverage"]] <- params$plotDensityCoverage
}
```

# Density {.tabset}
```{r, results='asis', echo = FALSE}
putInTabs(plots)
```

<!-- Figure 2 -->

# Reference CpGs covered {.tabset}
```{r, echo=FALSE}
plots <- list()
plots[["Count"]] <- params$plotCpGsCoveredCountBySampleAll
plots[["Proportion"]] <- params$plotCpGsCoveredFracBySampleAll
```

## By sample {.tabset}
```{r, results='asis', echo = FALSE}
putInTabs(plots, level = 3)
```

```{r, echo=FALSE}
plots <- list()
plots[["Count"]] <- params$plotCpGsCoveredCountByChrAll
plots[["Proportion"]] <- params$plotCpGsCoveredFracByChrAll
```

## By chromosome {.tabset}
```{r, results='asis', echo = FALSE}
putInTabs(plots, level = 3)
```

## By sample/chromosome {.tabset}
```{r, echo=FALSE}
plots <- list()
plots[["Count"]] <- params$plotCpGsCoveredCountBySample
plots[["Proportion"]] <- params$plotCpGsCoveredFracBySample
```

```{r, results='asis', echo = FALSE}
putInTabs(plots, level = 3)
```

```{r, echo=FALSE}
plots <- list()
plots[["Count"]] <- params$plotCpGsCoveredCountByChr
plots[["Proportion"]] <- params$plotCpGsCoveredFracByChr
```

## By chromosome/sample {.tabset}
```{r, results='asis', echo = FALSE}
putInTabs(plots, level = 3)
```

{- .leave-tabset}

<!-- Figure 3 -->

# Methylation {.tabset}
```{r, echo=FALSE}
plots <- list()
plots[["Whole genome"]] <- params$plotMethylationAll
plots[["By chromosome"]] <- params$plotMethylation
```

```{r, results='asis', echo = FALSE}
putInTabs(plots)
```

<!-- Figure 4 -->
`r if (!params$hasCov) {"<!--"}`

```{r, echo=FALSE}
plots <- list()
plots[["Whole genome"]] <- params$plotCoverageAll
plots[["By chromosome"]] <- params$plotCoverage
```

# Coverage {.tabset}
```{r, results='asis', echo = FALSE}
  putInTabs(plots)
```

`r if (!params$hasCov) {"-->"}`

End of report