---
  title: "04: DMR and Annotation"
author: "Andrew Lindsay, Reka Toth"
date: "`r Sys.Date()`"
output: 
  html_document:
  toc: true
toc_depth: 3
toc_float: true
self_contained: yes
highlight: pygments
vignette: >
  %\VignetteIndexEntry{scMethrix tutorial}
%\VignetteEngine{knitr::rmarkdown}
%\VignetteEncoding{UTF-8}
---


# Differential methylation calling

## DSS
A very important step of a WGBS data analysis is the differential methylation calling. During this step, we would like to identify changes between groups or conditions first on the site level. Then, we will need to identify larger regions (DMR-s, differentially methylated regions) affected by methylation changes that are more likely to represent functional alterations. `Methrix` doesn't have a DMR caller, therefore we will use `DSS`. 
At this stage, `DSS` is not yet able to work directly with `methrix`, therefore we transform our methrix object to `bsseq` for the sake of DMR calling. 

```{r message=FALSE, warning=FALSE}
if(!requireNamespace("DSS")) {
BiocManager::install("DSS")
}
library(DSS)
```

For more detailed description, options and recommendations on smoothing, please refer to the ´DSS´ [vignette](https://bioconductor.org/packages/release/bioc/vignettes/DSS/inst/doc/DSS.html)

```{r message=FALSE, warning=FALSE,}
bs_obj <- methrix::methrix2bsseq(meth)
```

```{r message=FALSE, warning=FALSE, include=FALSE}
dmlTest <- readRDS("DSS_dmltest.RDS")
```


```{r message=FALSE, warning=FALSE, echo=T, results='hide', eval=F}
dmlTest = DSS::DMLtest(bs_obj, group1=rownames(sample_anno)[sample_anno$Day=="D6"], group2=rownames(sample_anno)[sample_anno$Day=="D14"], 
                       smoothing = TRUE)
```

### Call DMRs and DMLs 


```{r warning=FALSE, include=FALSE}
if(!requireNamespace("kableExtra")) {
install.packages("kableExtra")
}
if(!requireNamespace("magrittr")) {
install.packages("magrittr")
}
library(kableExtra)
library(magrittr)
```


```{r results=FALSE}
dmls = DSS::callDML(dmlTest, p.threshold = 0.05)
dmrs = DSS::callDMR(dmlTest, p.threshold = 0.05)
```

**DMRs:**

```{r}
kable(dmrs) %>%
  kable_styling(bootstrap_options = "striped", full_width = F) %>%
  scroll_box(width = "100%", height = "200px")
```

**DMLs:**

```{r}
kable(dmls) %>%
  kable_styling(bootstrap_options = "striped", full_width = F) %>%
  scroll_box(width = "100%", height = "200px")
```

## DMR calling with `dmrseq`

A popular `R` package for DMR calling is `dmrseq`. For more detailed description of the possibilites, see the [vignette](http://bioconductor.org/packages/release/bioc/vignettes/dmrseq/inst/doc/dmrseq.html).

```{r message=FALSE, warning=FALSE, eval=FALSE}
if(!requireNamespace("dmrseq")) {
BiocManager::install("dmrseq")
}
library(dmrseq)
register(MulticoreParam(2))
testCovariate <- "Day"
regions <- dmrseq(bs=bs_obj[240001:260000,],
                  cutoff = 0.05,
                  testCovariate=testCovariate, verbose = T)
saveRDS(regions, "dmrseq_regions.RDS")
```

# Visualizing DMRs and candidate regions

## Visualizing a region 
We can visualize certain regions and/or DMRs using the [Gviz](https://bioconductor.org/packages/release/bioc/html/Gviz.html) package. The plotting function is included in the [best_practices_WGBS.Rmd](https://github.com/CompEpigen/WGBS_best_practice/blob/master/best_practices_WGBS.Rmd) file.

```{r message=FALSE, warning=FALSE, include=FALSE}
if(!requireNamespace("Gviz")) {
BiocManager::install("Gviz")
}
if(!requireNamespace("biomaRt")) {
BiocManager::install("biomaRt")
}
library(Gviz)
library(biomaRt)
```


```{r include=FALSE}
region_plot <- function(m = meth, region=candidate_region, mart=mart, genome=genome, groups=NULL, dmrs=NULL){
  candidate_region <- GenomicRanges::flank(region, width = 1000, both = TRUE)
## Annotation data from biomart
biomartTrack <- BiomartGeneRegionTrack(biomart = mart,
                                           chromosome=as.character(seqnames(candidate_region)),
                                           start=start(candidate_region),
                                           end=end(candidate_region), name="ENSEMBL")
# make it nice
displayPars(biomartTrack) <- list(background.title="transparent",fontcolor="black", 
                                  fontcolor.title="black", fontcolor.legend="black", col="black", 
                                  col.line="black", cex=3)
#Prepare methylation data
beta <- get_matrix(methrix::subset_methrix(m, regions = candidate_region), add_loci=TRUE)
beta$end <- beta$start+1
beta <- makeGRangesFromDataFrame(beta, keep.extra.columns = T)
#idiogram
itrack <- IdeogramTrack(genome=genome, chromosome=as.character(seqnames(candidate_region)),
            cex=1.2, fontcolor="black")
### Prepare tracks with methylation data
if (!is.null(groups)){
methylation_track <-DataTrack(beta,
                              lwd = 2,
                              cex = 20,
                              type=c("p", "smooth"),
                              span=0.5,
                              legend=TRUE,
                              groups=groups,
                              aggregateGroups=TRUE,
                              aggregation="mean",
                              na.rm=TRUE, 
                              name = "Methylation")
} else {
  methylation_track <-DataTrack(beta,
                              lwd = 2,
                              cex = 20,
                              type=c("p", "smooth"),
                              span=0.5,
                              legend=TRUE,
                              name = "Methylation")
  
}
methylation_track_hm <-DataTrack(beta,
                                 lwd = 2,
                                 cex = 10,
                                 type="heatmap",
                                 ncolor=100,
                                 name = "Methylation heatmap", 
                                 showSampleNames=TRUE, cex.sampleNames=0.6)
displayPars(methylation_track) <- list(fontcolor="black", fontcolor.title="black", fontcolor.legend="black", 
                                       fill="white", col.axis="black", cex.title=0.5, 
                                       background.title="transparent", cex.legend=0.8, cex=0.1)
displayPars(methylation_track_hm) <- list(fontcolor="black", fontcolor.title="black", fontcolor.legend="black", 
                                           col.axis="black", cex.title =0.5, col.sampleNames="black", cex.sampleNames=0.5,
                                          background.title="transparent", cex.legend=1.2)
gtrack <- GenomeAxisTrack(fontcolor="black", col="black")
 
###annotation of DMRs, if needed
if (!is.null(dmrs)){
#dmr_data <-  makeGRangesFromDataFrame(dmrs, keep.extra.columns = T)
DMR_track <- HighlightTrack(trackList=list(biomartTrack), start=start(dmrs), width=width(dmrs), chromosome=as.character(seqnames(candidate_region)))
DMR_track_2 <- HighlightTrack(trackList=list(methylation_track), start=start(dmrs), width=width(dmrs), chromosome=as.character(seqnames(candidate_region)))
displayPars(DMR_track) <- list(col="grey50", fill="lightgrey", alpha=0.50)
displayPars(DMR_track_2) <- list(col="grey50", fill="lightgrey", alpha=0.50)
} 
## plot the datatracks
if (is.null(dmrs)){
  tracklist <- list(itrack, gtrack, biomartTrack, methylation_track_hm, methylation_track)
 } else {
  tracklist <-  list(itrack, gtrack, DMR_track, methylation_track_hm, DMR_track_2)}
                   
try(plotTracks(tracklist, 
                      sizes=c(0.5,1.5,1,3,6),
                    from = start(candidate_region),
                    to = end(candidate_region),
                    transcriptAnnotation="symbol",
                    silent = T))
}
```


```{r}
dmrs <- makeGRangesFromDataFrame(dmrs, keep.extra.columns = T)
```

```{r fig.height=4, fig.width=6, eval=F}
genome <- "hg19"
mart <- useMart(biomart="ENSEMBL_MART_ENSEMBL", host="feb2014.archive.ensembl.org", dataset="hsapiens_gene_ensembl")
i <- 5
#candidate_region <- GRanges(paste0(seqlevels(dmrs)[i], ":", start(dmrs)[i], "-", end(dmrs)[i]))
candidate_region <- dmrs[i]
region_plot(m = meth, region=candidate_region, mart=mart, genome=genome, groups=meth$Day, dmrs=dmrs)
#the region_plot function is available in the best_practices.Rmd file. 
```

## Heatmap of DMRs

```{r fig.height=4, fig.width=5, message=FALSE, warning=FALSE}
heatmap_data <- as.data.frame(methrix::get_region_summary(meth, makeGRangesFromDataFrame(dmrs)))
heatmap_data <- heatmap_data[complete.cases(heatmap_data),]
rownames(heatmap_data) <- paste0(heatmap_data$chr,":", heatmap_data$start, "-", heatmap_data$end)
pheatmap::pheatmap(heatmap_data[,-(1:5)], show_rownames = FALSE, annotation_col = as.data.frame(meth@colData), 
                   fontsize = 8)
```

## Number and size of DMRs 

```{r message=FALSE, warning=FALSE}
if(!requireNamespace("plotly")) {
install.packages("plotly")
}
if(!requireNamespace("ggplot2")) {
install.packages("ggplot2")
}
if(!requireNamespace("scales")) {
install.packages("scales")
}
library(plotly)
library(ggplot2)
library(scales)
```

```{r fig.height=3, fig.width=5}
count <- c("Higher methylation in day 14" = length(dmrs[dmrs$diff.Methy<0,]),
           "Higher methylation in day 6" = length(dmrs[dmrs$diff.Methy>0,]))
length <- c("Higher methylation in day 14" = sum(width(dmrs[dmrs$diff.Methy<0,])), 
            "Higher methylation in day 6" = sum(width(dmrs[dmrs$diff.Methy<0,]))) 
data <- data.frame(Direction=c("Gain", "Loss"), Count=count, Length=length)
   g <- ggplot(data=data)+geom_col(aes(x=factor(1), y=Count, fill=Direction), position = "dodge")+theme_bw()+theme(axis.title.x = element_blank(), axis.text.x = element_blank())+scale_fill_brewer(palette = "Dark2")+ggtitle("Number of differentially methylated regions")+scale_y_continuous(labels = comma)
ggplotly(g)
   p <- ggplot(data=data)+geom_col(aes(x=factor(1), y=Length, fill=Direction), position = "dodge")+theme_bw()+theme(axis.title.x = element_blank(), axis.text.x = element_blank())+scale_fill_brewer(palette = "Dark2")+ggtitle("Length of differentially methylated regions")+scale_y_continuous(labels = comma)
ggplotly(p)
   
```



# Region annotation 

To describe and interpret the differentially methylated regions, we can use the `ChIPseeker` package. The location of the DMRs can already give us a hint the involved processes, but once we assign the regions to genes, a pathway enrichment analysis is also possible. For additional details, see the [vignette](https://bioconductor.org/packages/release/bioc/vignettes/ChIPseeker/inst/doc/ChIPseeker.html) of `ChIPseeker`:  


```{r message=FALSE, warning=FALSE}
## loading packages
if(!requireNamespace("ChIPseeker")) {
BiocManager::install("ChIPseeker")
}
if(!requireNamespace("TxDb.Hsapiens.UCSC.hg19.knownGene")) {
BiocManager::install("TxDb.Hsapiens.UCSC.hg19.knownGene")
}
if(!requireNamespace("ReactomePA")) {
BiocManager::install("ReactomePA")
}
if(!requireNamespace("clusterProfiler")) {
BiocManager::install("clusterProfiler")
}
library(ChIPseeker)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
library(ReactomePA)
library(clusterProfiler)
```

```{r fig.height=3, fig.width=5}
 
  peakAnnoList <- lapply(list("Higher in day 14"=dmrs[dmrs$diff.Methy<0,], "Higher in day 6" = dmrs[dmrs$diff.Methy>0,]), annotatePeak, TxDb=txdb, tssRegion=c(-3000, 3000), verbose=FALSE)
plotAnnoBar(peakAnnoList)
plotDistToTSS(peakAnnoList)
genes <- lapply(list("Higher in day 14"=dmrs[dmrs$diff.Methy<0,], "Higher in day 6" = dmrs[1:3,]), seq2gene, tssRegion = c(-1000, 1000), flankDistance = 3000, TxDb=txdb)
names(genes) <- gsub(" ", "\n", names(genes))
comppthw <- compareCluster(geneCluster   = genes,
                         fun           = "enrichPathway",
                         pvalueCutoff  = 0.05,
                         pAdjustMethod = "BH")
dotplot(comppthw, font.size = 8)
gc()
```
