---
title: "01: Reading in BedGraph data"
author: "Andrew Lindsay, Reka Toth"
date: "`r Sys.Date()`"
vignette: >
  %\VignetteIndexEntry{01: Reading in BedGraph data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, knitr-setup, include=FALSE}
library(knitr)

knitr::opts_chunk$set(echo = TRUE, 
                      collapse = TRUE,
                      results= "hold",
                      eval=FALSE,
                      comment = "#>")

hook_output <- knit_hooks$get("output")
knit_hooks$set(output = function(x, options) {
   lines <- options$output.lines
   if (is.null(lines)) {
     return(hook_output(x, options))  # pass to default hook
   }
   x <- unlist(strsplit(x, "\n"))
   more <- "..."
   if (length(lines)==1) {        # first n lines
     if (length(x) > lines) {
       # truncate the output, but add ....
       x <- c(head(x, lines), more)
     }
   } else {
     x <- c(more, x[lines], more)
   }
   # paste these lines together
   x <- paste(c(x, ""), collapse = "\n")
   hook_output(x, options)
 })
```

# Load the library

```{r, echo=T, results='hide'}
#Load library
library(scMethrix)
```

# Download the data

The example data analysis presented here uses the following publicly available whole genome bisulfite sequencing data from GEO: [GSE56879](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE56879)

> Smallwood SA, Lee HJ, Angermueller C, Krueger F et al. Single-cell genome-wide bisulfite sequencing for assessing epigenetic heterogeneity. Nat Methods 2014 Aug; 11(8):817-820. [PMID: 25042786](https://pubmed.ncbi.nlm.nih.gov/25042786/)

In this dataset, mouse embryonic stem cells (ESCs) have been grown in either serum or 2i medium. 2i medium is known to prevent differentiation of ESCs as well as induce global hypomethylation ([Sim et al, 2017](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5425728/)), so the methylome should show drastic differences. In this example, 32 samples will be taken; 12 grown in 2i, and the rest in serum.

The data can be directly downloaded from GEO, using the `GEOquery` package. The files will be downloaded into the working directory.

```{r, echo=T, results='hide'}
if(!requireNamespace("GEOquery")) {
  BiocManager::install("GEOquery")
}
library(GEOquery)  
```

```{r, message=FALSE, warning=FALSE}
data_dir = paste0("D:/Git/sampleData/Vignette.GSE56879/")
dir.create(path = data_dir, showWarnings = FALSE, recursive = TRUE)
```

```{r import-data, cache=TRUE, include=FALSE}
# Download the dataset
gsm_list <- c(paste0("GSM",1370535+0:11),  paste0("GSM",1370555+0:19))

bed_files <- sapply(gsm_list, function(gsm){
GEOquery::getGEOSuppFiles(GEO = gsm, baseDir = data_dir, makeDirectory = FALSE, filter_regex = ".*.cov.txt.gz")
})

# Rename to bedgraph format
bed_files <- list.files(path = data_dir,  full.names = TRUE, pattern = ".*.cov.txt.gz")
file.rename(bed_files,gsub("(.*GSM.*?)_.*","\\1.bedgraph.gz",bed_files))
bed_files <- list.files(path = data_dir,  full.names = TRUE, pattern = ".*.bedgraph.gz")
```

```{r}
head(basename(bed_files))
```

Please note: The sample name in the `scMethrix` object is generated from the input file name minus the extension (e.g., `GSM1370535.bedgraph` becomes `GSM1370535`. As [`data.table::fread`](https://www.rdocumentation.org/packages/data.table/versions/1.14.2/topics/fread) is used, the `.gz` and `.bz2` extension will be dropped automatically as well (e.g., `GSM1370535.bedgraph.gz` becomes `GSM1370535`). See `get_sample_name()` for additional explanation.

------------------------------------------------------------------------

# Processing

## CpG annotation

As a first step, we need a list of CpG sites in the respective genome. The CpG can be found in the respective [`BSgenome`](https://bioconductor.org/packages/release/bioc/html/BSgenome.html) annotation package. If the reference CpGs (`ref_cpgs`) are not supplied, the `read_beds()` function is able to extract CpG sites from the raw input data. This is significantly more efficient for experiments where few CpGs are read (e.g., [RRBS](https://pubmed.ncbi.nlm.nih.gov/16224102/)), but there will be no automatic filtering of erroneous CpG sites. The user can also manually filter the `ref_cpgs` before usage to improve performance.

```{r, include = FALSE}
#Genome of your preference to work with
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
library(BiocManager)
if(!requireNamespace("BSgenome.Mmusculus.UCSC.mm10")) {
  BiocManager::install("BSgenome.Mmusculus.UCSC.mm10")
}
library(BSgenome.Mmusculus.UCSC.mm10) 
```

```{r, ref-cpgs, cache=TRUE}
mm10_cpgs <- suppressWarnings(scMethrix::extract_CpGs(genome = "BSgenome.Mmusculus.UCSC.mm10"))
head(mm10_cpgs)
```

## Sample annotation

An annotation table for sample data is also necessary to perform some analyses. The data will be added to the `scMethrix` object in the `colData` slot. Each row of the table represents a single sample. Sample names are taken from the input file names, and stored as row.names. Subsequent information about each sample (e.g., cell type, alternate IDs, sample source, etc) can be stored as columns within this `data.frame`. To access this data, see [Saving experiment metadata].

For inputted colData, the row name must match the corresponding sample name. See [Download the data](#download) for how sample names are generated. The row order of the inputted colData is not important as it will be added as a left join (`merge(all.x=T)`) to the sample names. Hence, the inputted colData can contain samples that are not present in the inputted files, and their respective colData will not be included in the experiment.

For this example, the desired colData can be easily extracted from the experiment's [SOFT file in GEO](https://ftp.ncbi.nlm.nih.gov/geo/series/GSE56nnn/GSE56879/soft/).

```{r, colData, cache=TRUE, echo=T, results='hide'}
soft <- GEOquery::getGEOfile("GSE56879")
soft <- GEOquery::getGEO(filename=soft)

sample_data <- sapply(names(soft@gsms), function(gsm) {
  soft@gsms[[gsm]]@header$title
})

sample_data <- gsub("_.*","",sample_data)
#sample_data <- sample_data[which(names(soft@gsms) %in% gsm_list)]
sample_data <- data.frame(row.names = names(sample_data), Medium = sample_data)
```

```{r}
head(sample_data)
table(sample_data)
```

## Reading `BedGraph` files

The `read_beds()` function is a versatile file reader intended to import `BedGraph` files generated virtually by any sort of methylation calling program. It requires user to provide indices for chromosome names, start position and other required fields. There are also presets available to import `BedGraph` files from most common methylation callers such as `Bismark`, `MethylDackel`, and `MethylcTools`. In this case, there is no need to define `chr_idx`, `start_idx`, or other `idx` arguments as the function will automatically assign them.

To know the exact parameters of our input dataset, it might worth to take a look at one of the files, to see the column order.

```{r}
res <- data.table::fread(bed_files[1])
head(res)
```

Either given by the data source, or from looking at the values, it is obvious the respective columns are for chromosome (`chr`), start position (`start`), end position (`end`), beta value (`beta`), \# of methylated sites (`M`), and \# of unmethylated sites (`U`). The column index for each will be specified as arguments in `read_beds()`.

The `read_beds()` function will aggregate and process the inputted data files. Once the process is complete, it returns an object of class `scMethrix()`, which extends the `SingleCellExperiment()` class. `scMethrix()` objects contain methylation and (optionally) coverage data as either in-memory matrices or as on-disk HDF5 arrays, as well as our phenotype data and other basic metadata. This object can be passed to all downstream functions for various analysis. For further details on the data structure, see the `SingleCellExperiment()` package:

A more detailed description of the `read_beds()` function and potential pitfalls in usage can be found [here](x07_read_beds.html)

```{r, read_beds}
scm <- scMethrix::read_beds(
  files = bed_files,
  ref_cpgs = mm10_cpgs,
  h5 = TRUE,
  h5_dir = "D:\\Git\\sampleData\\Vignette.GSE56879\\h5",
  chr_idx = 1,
  start_idx = 2,
  end_idx = 3,
  beta_idx = 4,
  M_idx = 5,
  U_idx = 6,
  stranded = FALSE,
  zero_based = FALSE, 
  colData = sample_data
)
```

```{r}
scm
```

For reading other file types (e.g. `IDAT` or `bigWig`), see [other input formats](x08_read_other.html)

# Experiment data

## Assay matrices

After the experiment is created, an assay matrix can be generated. The methylation and coverage matrices can be directly obtained with the functions `score()` and `counts()`, respectively, while any matrix can be accessed via `get_matrix(scm, assay="ASSAY_NAME")`.

```{r}
mtx <- score(scm)
mtx[10,6]
```

```{r}
mtx <- get_matrix(scm, assay="score")
mtx[10,6]
```

Genomic loci can also be added (`add_loci`), be formatted as a `GenomicRanges` object, and ordered by variance (`order_by_sd`), if desired.

```{r}
# Add genomic loci, put in a GRanges object, and order by SD
mtx <- get_matrix(scm, assay="score", add_loci = TRUE, in_granges=TRUE, order_by_sd = TRUE)
head(mtx)
```

## Metadata

`scMethrix` also stores experiment metadata. Depending on the type of data, it is stored in different places:

1.  **Experiment-wide metadata:** General information is stored in `metadata()` as a named `list`. This can be information like genome accession, GEO IDs, or authors.

```{r indent = "     "}
metadata(scm)
```

2.  **Range metadata:** Information about ranges is stored as a `data.frame` within the `rowRanges()` sub-object inside the `scMethrix` object, and can be accessed by `rowData()`. This can include gene names, region IDs, count numbers (if binned). In this example, a random ID is generate for each CpG site.

```{r, indent = "     "}
head(rowData(scm))
```

3.  **Sample metadata:** Information about samples is stored as a `data.frame` in `colData()`. This can include alternate samples IDs, run IDs, tissue types, or anything associated with individual samples. In this example, only the medium used for growing each sample is stored.

```{r, indent = "     "}
head(colData(scm))
```

Certain operations will cause metadata to be destroyed. For range metadata, `bin_scMethrix()` will discard row metadata (except for CpG count), as will `collapse_samples()` for sample metadata (except for sample names). It is advised to extract this data from `rowData()` and `colData()`, respectively, and store it within `metadata()` instead, or aggregate and store the data for each respective region in their respective container.


[scMethrix]: https://compepigen.github.io/scMethrix/reference/scMethrix.html "`scMethrix`"
