---
title: "04: Dimensionality reduction and clustering"
author: "Andrew Lindsay, Reka Toth"
date: "`r Sys.Date()`"
output: 
  html_document:
  toc: true
toc_depth: 3
toc_float: true
self_contained: yes
highlight: pygments
vignette: >
  %\VignetteIndexEntry{04: Dimensionality reduction and clustering}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Dimensionality reduction 

`scMethrix` offers principal component analysis (PCA), uniform manifold approximation and projection (UMAP), and t-distributed stochastic neighbor embedding (tSNE) to conduct and visualize dimensionality reduction. The resulting reduction is stored in the scMethrix object, and can be accessed via reducedDim(). Typically, the number of CpG sites are reduced via `reduce_scMethrix` to ensure the calculations remain feasible. 

```{r fig.height=3, fig.width=4, eval=FALSE}
scm.red <- reduce_scMethrix(scm, var = "top", n_cpg = 1000)
scm <- scMethrix::dim_red_scMethrix(scm,type="PCA", top_var = 1000, n_components = 20)
scMethrix::plot_dim_red(scm,type="PCA")
```

```{r fig.height=3, fig.width=4, eval=FALSE}
scm <- scMethrix::dim_red_scMethrix(scm,type="UMAP",top_var = 1000, n_neighbors = 3)
scMethrix::plot_dim_red(scm,type="UMAP")
```

```{r fig.height=3, fig.width=4, eval=FALSE}
scm <- scMethrix::dim_red_scMethrix(scm,type="tSNE",top_var = 1000, n_components = 20)
scMethrix::plot_dim_red(scm,type="tSNE")
```

At visualization, we can provide the `scMethrix` object to allow color or shape annotation of groups or samples.  

```{r fig.height=3, fig.width=4, eval=FALSE}
scMethrix::plot_pca(scm, col_anno = "Day", shape_anno = "Replicate")
```

`scMethrix` offers the possibility of region based filtering. With this option, selected regions (e.g. promoters) can be visualized. We will use the `AnnotationHub` package to assess basic annotation categories. 

```{r message=FALSE, warning=FALSE, eval=FALSE}
if(!requireNamespace("AnnotationHub")) 
  BiocManager::install("AnnotationHub", update = F)
```

```{r message=FALSE, warning=FALSE, eval=FALSE}
library("AnnotationHub")
ah = AnnotationHub()
qhs = query(ah, c("RefSeq", "Mus musculus", "mm10"))
genes = qhs[[1]]
promotors = promoters(genes)
```

```{r fig.height=3, fig.width=4, message=FALSE, warning=FALSE, eval=FALSE}
meth.subset <- subset_methrix(meth,regions = promotors)
meth <- scMethrix::dim_red_scMethrix(meth.subset,type="PCA",top_var = 10000, n_components = 20)
plot_dim_red(meth,type="PCA")
```

# Clustering data
Unlike bulk sequencing, where the measurements are averaged over many cells, single-cell sequencing allows the identification of cell type for individual cells. This is a common step in many single-cell expression workflows. Multiple algorithms and distance metrics are included in the package for maximum flexibility. 

## Distance metrics
Most algorithms operate via a distance metric, an all-by-all matrix where each sample is numerically compared to each other sample. Samples with few methylated CpG sites in common will have a large distance metric, where as two samples with identical methylation will have a distance metric of zero. 

Available distance metrics include `pearson`, `spearman`, and `tau` (all via `bioDist`); and `euclidean`, `maximum`, `manhattan`, `canberra`, `binary`, and `minkowski` (all via `stats::dist`). 

```{r, eval=F}
scMethrix::get_distance_matrix(scm, assay = "score" , type = "pearson")
```

Additional distance metrics can be used via an arbitrary function input. The input must the assay matrix and output must be an all-by-all matrix filled with distance values for each respective pair. It will be internally cast with `as.dist()` before computation.

```{r, eval=F}
fun <- function(mtx) dist(mtx, method = "euclidean")
scMethrix::get_distance_matrix(scm, assay = "score" , type = fun)
```

## Clustering
Clustering algorithms include `hierarchical` (via `stats::hclust`), `partition` (via `stats::kmeans`), and `model`-based (via `mclust::Mclust`). Distance metrics are not used for model-based clustering. The identified clusters will be stored in the `colData` slot of the output `scMethrix` object.

```{r, eval=F}
scm.cluster <- scMethrix::cluster_scMethrix(scm, assay = "score", dist = dist, n_clusters = 2, colname = "Heirarchical", type="hierarchical")
colData(scm.cluster)
```

Like the distance metric, an arbitrary clustering algorithm can be used. It must accept a `dist` object, and return a `data.frame` with two columns named "Sample" and "Cluster". The column "Cluster" will be renamed by the value in `colname` before returning, if given. 

```{r, eval=F}
fun <- function (dist) {
  fit <- stats::hclust(dist, method="ward.D")
  fit <- stats::cutree(fit, k=ncol(dist)/2)
  colData <- data.frame(Sample = names(fit), Cluster = fit)
  colData
}

scm.cluster <- scMethrix::cluster_scMethrix(scm, assay = "score", dist = dist, n_clusters = 2, colname = "Heir", type=fun)
colData(scm.cluster)
```

# Visualization

For visualizing the clustering, the generated colData can be used for annotation in dimensionality reduction plots.

```{r, eval=F}
scm <- scMethrix::dim_red_scMethrix(scm,type="PCA",top_var = 10000)
plot_dim_red(scm,type="PCA",shape_anno = "Cluster",color_anno="Cluster")
```