<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ scMethrix</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMethrix</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.0001</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/CompEpigen/scMethrix/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-bed-input-functions" class="hasAnchor"><a href="#section-bed-input-functions" class="anchor"></a>BED input functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="read_beds.html">read_beds()</a></code> </p>
        </td>
        <td><p>Versatile BedGraph reader.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_index.html">read_index()</a></code> </p>
        </td>
        <td><p>Parse BED files for unique genomic coordinates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_bed_by_index.html">read_bed_by_index()</a></code> </p>
        </td>
        <td><p>Parses BED files for methylation values using previously generated index genomic coordinates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_hdf5_data.html">read_hdf5_data()</a></code> </p>
        </td>
        <td><p>Writes values from input BED files into an in-disk <code>HDF5Array</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_mem_data.html">read_mem_data()</a></code> </p>
        </td>
        <td><p>Writes values from input BED files into an in-memory <code><a href='https://rdrr.io/r/base/matrix.html'>matrix</a></code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subset_ref_cpgs.html">subset_ref_cpgs()</a></code> </p>
        </td>
        <td><p>Subsets a given list of CpGs by another list of CpGs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_sample_name.html">get_sample_name()</a></code> </p>
        </td>
        <td><p>Returns sample name derived from the input file name</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-base-object-and-data" class="hasAnchor"><a href="#section-base-object-and-data" class="anchor"></a>Base object and data</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="scMethrix-class.html">scMethrix-class</a></code> </p>
        </td>
        <td><p>Class scMethrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scMethrix_data.html">scMethrix_data</a></code> </p>
        </td>
        <td><p>WGBS for single cell methylome of murine neuronal cells</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-object-getters" class="hasAnchor"><a href="#section-object-getters" class="anchor"></a>Object getters</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="is_h5.html">is_h5()</a></code> </p>
        </td>
        <td><p>Checks if <code>scMethrix</code> object is an HDF5 object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="has_cov.html">has_cov()</a></code> </p>
        </td>
        <td><p>Checks if <code>scMethrix</code> object has a coverage matrix</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-scmethrix-base-operations" class="hasAnchor"><a href="#section-scmethrix-base-operations" class="anchor"></a>scMethrix base operations</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-save-load-and-conversion" class="hasAnchor"><a href="#section-save-load-and-conversion" class="anchor"></a>Save/load and conversion</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="save_HDF5_scMethrix.html">save_HDF5_scMethrix()</a></code> </p>
        </td>
        <td><p>Saves an HDF5 <code>scMethrix</code> object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="load_HDF5_scMethrix.html">load_HDF5_scMethrix()</a></code> </p>
        </td>
        <td><p>Loads HDF5 <code>scMethrix</code> object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convert_HDF5_scMethrix.html">convert_HDF5_scMethrix()</a></code> </p>
        </td>
        <td><p>Converts HDF5 <code>scMethrix</code> object to an in-memory <code>scMethrix</code> object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convert_scMethrix.html">convert_scMethrix()</a></code> </p>
        </td>
        <td><p>Converts an in-memory <code>scMethrix</code> to an HDF5 <code>scMethrix</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convert_to_methrix.html">convert_to_methrix()</a></code> </p>
        </td>
        <td><p>Converts an <code>scMethrix</code> object to methrix object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="export_bed.html">export_bed()</a></code> </p>
        </td>
        <td><p>Exports all samples in an <code>scMethrix</code> objects into individual bedgraph files</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-assay-retrieval" class="hasAnchor"><a href="#section-assay-retrieval" class="anchor"></a>Assay retrieval</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="remove_assay.html">remove_assay()</a></code> </p>
        </td>
        <td><p>Removes an assay from an <code>scMethrix</code> object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_matrix.html">get_matrix()</a></code> </p>
        </td>
        <td><p>Extract assays from an <code>scMethrix</code> object</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-operations" class="hasAnchor"><a href="#section-operations" class="anchor"></a>Operations</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="merge_scMethrix.html">merge_scMethrix()</a></code> </p>
        </td>
        <td><p>Merges two <code>scMethrix</code> objects by <code>row</code> or <code>col</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subset_scMethrix.html">subset_scMethrix()</a></code> </p>
        </td>
        <td><p>Subsets an <code>scMethrix</code> object based on <code>regions</code>, <code>contigs</code> and/or <code>samples</code>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mask_scMethrix.html">mask_scMethrix()</a></code> </p>
        </td>
        <td><p>Masks high or low coverage <code>count</code> or <code>cell</code> count</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="coverage_filter.html">coverage_filter()</a></code> </p>
        </td>
        <td><p>Filter matrices by coverage</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="remove_uncovered.html">remove_uncovered()</a></code> </p>
        </td>
        <td><p>Remove loci that are uncovered across all samples</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-statistics" class="hasAnchor"><a href="#section-statistics" class="anchor"></a>Statistics</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="get_stats.html">get_stats()</a></code> </p>
        </td>
        <td><p>Estimate descriptive statistics for each sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_region_summary.html">get_region_summary()</a></code> </p>
        </td>
        <td><p>Extracts and summarizes methylation or coverage info by regions of interest</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_metadata_stats.html">get_metadata_stats()</a></code> </p>
        </td>
        <td><p>Adds descriptive statistics to metadata columns in an <code>scMethrix</code> object.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-scmethrix-transformations" class="hasAnchor"><a href="#section-scmethrix-transformations" class="anchor"></a>scMethrix transformations</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-assay-transform" class="hasAnchor"><a href="#section-assay-transform" class="anchor"></a>Assay transform</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="transform_assay.html">transform_assay()</a></code> </p>
        </td>
        <td><p>Transforms an assay in an <code>scMethrix</code> object.</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-imputation" class="hasAnchor"><a href="#section-imputation" class="anchor"></a>Imputation</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="impute_by_iPCA.html">impute_by_iPCA()</a></code> </p>
        </td>
        <td><p>Imputes the an <code>scMethrix</code> object using a iterative PCA model</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="impute_by_melissa.html">impute_by_melissa()</a></code> </p>
        </td>
        <td><p>Imputes the NA values of a <code>scMethrix</code> object.</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-dimensionality-reduction" class="hasAnchor"><a href="#section-dimensionality-reduction" class="anchor"></a>Dimensionality Reduction</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="pca_scMethrix.html">pca_scMethrix()</a></code> </p>
        </td>
        <td><p>Principal Component Analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="umap_scMethrix.html">umap_scMethrix()</a></code> </p>
        </td>
        <td><p>Generates UMAP for scMethrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tsne_scMethrix.html">tsne_scMethrix()</a></code> </p>
        </td>
        <td><p>Generates tSNE for scMethrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reduce_matrix.html">reduce_matrix()</a></code> </p>
        </td>
        <td><p>Reduces a matrix</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-plotting" class="hasAnchor"><a href="#section-plotting" class="anchor"></a>Plotting</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="prepare_plot_data.html">prepare_plot_data()</a></code> </p>
        </td>
        <td><p>Format <code>scMethrix</code> matrix to long form data for plotting</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_violin.html">plot_violin()</a></code> </p>
        </td>
        <td><p>Violin Plot for \(\beta\)-Values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_density.html">plot_density()</a></code> </p>
        </td>
        <td><p>Density Plot of \(\beta\)-Values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_coverage.html">plot_coverage()</a></code> </p>
        </td>
        <td><p>Coverage QC Plots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_stats.html">plot_stats()</a></code> </p>
        </td>
        <td><p>Plot descriptive statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_dim_red.html">plot_dim_red()</a></code> </p>
        </td>
        <td><p>Plot PCA results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_pca.html">plot_pca()</a></code> </p>
        </td>
        <td><p>Plot PCA results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_umap.html">plot_umap()</a></code> </p>
        </td>
        <td><p>Plot UMAP results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_tsne.html">plot_tsne()</a></code> </p>
        </td>
        <td><p>Plot tSNE results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_palette.html">get_palette()</a></code> </p>
        </td>
        <td><p>Getter for plot palette colors</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-misc-functions" class="hasAnchor"><a href="#section-misc-functions" class="anchor"></a>Misc functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-transforms" class="hasAnchor"><a href="#section-transforms" class="anchor"></a>Transforms</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="binarize.html">binarize()</a></code> </p>
        </td>
        <td><p>Binarize an input value based on a <code>threshold</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="binarize2.html">binarize2()</a></code> </p>
        </td>
        <td><p>Binarize an input value based on a <code>threshold</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bin_scMethrix.html">bin_scMethrix()</a></code> </p>
        </td>
        <td><p>Bins the ranges of an <code>scMethrix</code> object.</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-granges-list" class="hasAnchor"><a href="#section-granges-list" class="anchor"></a>GRanges/List</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="bin_granges.html">bin_granges()</a></code> </p>
        </td>
        <td><p>Bins each region in a <code>GRanges</code> object into bins of specified <code>bin_size</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cast_granges.html">cast_granges()</a></code> </p>
        </td>
        <td><p>Casts genomic regions into <code>GRanges</code> format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="split_granges.html">split_granges()</a></code> </p>
        </td>
        <td><p>Splits a <code>GRanges</code> object by <code>chunks</code>, <code>percent</code> or <code>size</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="split_vector.html">split_vector()</a></code> </p>
        </td>
        <td><p>Splits a vector into subvectors by <code>chunk</code> or <code>size</code></p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-timer" class="hasAnchor"><a href="#section-timer" class="anchor"></a>Timer</h3>
          <p class="section-desc"></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="start_time.html">start_time()</a></code> </p>
        </td>
        <td><p>Starts an internal stopwatch</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="split_time.html">split_time()</a></code> </p>
        </td>
        <td><p>Outputs the split/lap/iteration time</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stop_time.html">stop_time()</a></code> </p>
        </td>
        <td><p>Stops an internal stopwatch and outputs overall time</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Lindsay Andrew.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


