<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>02: QC and reports • scMethrix</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="02: QC and reports">
<meta property="og:description" content="scMethrix">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMethrix</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.0001</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_load_data.html">01: Reading in the bedgraph data</a>
    </li>
    <li>
      <a href="../articles/02_qc.html">02: QC and reports</a>
    </li>
    <li>
      <a href="../articles/03_transforms.html">03: Transformations and Imputations</a>
    </li>
    <li>
      <a href="../articles/04_clustering.html">04: Clustering and metacells</a>
    </li>
    <li>
      <a href="../articles/05_hdf5.html">05: Working with large cohorts</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/CompEpigen/scMethrix/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="02_qc_files/header-attrs-2.10/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>02: QC and reports</h1>
                        <h4 class="author">Andrew Lindsay, Reka Toth</h4>
            
            <h4 class="date">2021-09-09</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/CompEpigen/scMethrix/blob/master/vignettes/02_qc.Rmd"><code>vignettes/02_qc.Rmd</code></a></small>
      <div class="hidden name"><code>02_qc.Rmd</code></div>

    </div>

    
    
<div id="filtering" class="section level1">
<h1 class="hasAnchor">
<a href="#filtering" class="anchor"></a>Filtering</h1>
<div id="coverage-based-filtering" class="section level2">
<h2 class="hasAnchor">
<a href="#coverage-based-filtering" class="anchor"></a>Coverage based filtering</h2>
<p>To ensure the high quality of our dataset, the sites with very low (untrustworthy methylation level) or high coverage (technical problem) should not be used. We can mask these CpG sites. Please note, that the DSS we were using for Differential Methylation Region (DMR) calling, doesn’t need the removal of the lowly covered sites, because it takes it into account by analysis. However, using a not too restrictive threshold won’t interfere with DMR calling.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meth</span> <span class="op">&lt;-</span> <span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="../reference/mask_by_coverage.html">mask_by_coverage</a></span><span class="op">(</span><span class="va">meth</span>, low_threshold <span class="op">=</span> <span class="fl">2</span>, avg_threshold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> </code></pre></div>
<p>Later we can remove those sites that are not covered by any of the samples.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meth</span> <span class="op">&lt;-</span> <span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="../reference/remove_uncovered.html">remove_uncovered</a></span><span class="op">(</span><span class="va">meth</span><span class="op">)</span> </code></pre></div>
<p>To reduce the data for future imputation/binning/clustering, we might want to remove non-variable CpG sites. The function <code>mask_by_variance</code> will calculate the variance of each CpG site and mask it if below a specified threshold.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meth</span> <span class="op">&lt;-</span> <span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="../reference/mask_by_variance.html">mask_by_variance</a></span><span class="op">(</span><span class="va">meth</span>, low_threshold <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></code></pre></div>
</div>
<div id="snp-filtering" class="section level2">
<h2 class="hasAnchor">
<a href="#snp-filtering" class="anchor"></a>SNP filtering</h2>
<p>SNPs, mostly the C &gt; T mutations can disrupt methylation calling. Therefore it is essential to remove CpG sites that overlap with common variants, if we have variation in our study population. For example working with human, unpaired data, e.g. treated vs. untreated groups. During filtering, we can select the population of interest and the minimum minor allele frequency (MAF) of the SNPs. SNP filtering is currently implemented for mm10, hg19 and hg38 assemblies.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"GenomicScores"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"GenomicScores"</span><span class="op">)</span>
<span class="op">}</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GenomicScores</span><span class="op">)</span> </code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meth</span> <span class="op">&lt;-</span> <span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/methrix/man/remove_snps.html">remove_snps</a></span><span class="op">(</span><span class="va">meth</span>, keep <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<ul>
<li>There is a possibility to keep the sites possibly overlapping with SNPs. We can use this data for detecting possible sample mismatches. For this, we need to plot the most variable sites (probably the ones that are variable in our study population) and plot them on a heatmap.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"pheatmap"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"pheatmap"</span><span class="op">)</span>
<span class="va">snps</span> <span class="op">&lt;-</span> <span class="va">meth</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>
<span class="va">meth</span> <span class="op">&lt;-</span> <span class="va">meth</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_matrix.html">get_matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/methrix/man/order_by_sd.html">order_by_sd</a></span><span class="op">(</span><span class="va">snps</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="fl">5000</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">snps</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<ul>
<li>SNP removal in mouse data: Usually, there is no need to remove SNPs from mouse data, since for many experiments, the same strain of mice is used. However, there are cases, when different crosses of strains are used, for which, the level of heterozygosity might be different. SNP data for mice can be downloaded from here: <a href="https://www.sanger.ac.uk/science/data/mouse-genomes-project" class="uri">https://www.sanger.ac.uk/science/data/mouse-genomes-project</a>. The SNPs can be filtered out using the <code>subset_scMethrix</code> function.</li>
</ul>
</div>
</div>
<div id="visualization-and-qc-after-filtering" class="section level1">
<h1 class="hasAnchor">
<a href="#visualization-and-qc-after-filtering" class="anchor"></a>Visualization and QC after filtering</h1>
<div id="scmethrix-report" class="section level2">
<h2 class="hasAnchor">
<a href="#scmethrix-report" class="anchor"></a>scMethrix report</h2>
<p>After filtering, it worth running a report again, to see if any of the samples were so lowly covered that it warrants an action (e.g. removal of the sample)</p>
<p><strong>Important:</strong> Don’t forget to set the <code>recal_stats</code> to <code>TRUE</code>, since the object changed since reading in. The output directory has to be different from last time, to avoid using the intermediate files calculated during the last run.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">methrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/methrix/man/methrix_report.html">methrix_report</a></span><span class="op">(</span>meth <span class="op">=</span> <span class="va">meth</span>, output_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span><span class="op">(</span><span class="op">)</span>, recal_stats <span class="op">=</span> <span class="cn">TRUE</span>, prefix<span class="op">=</span><span class="st">"processed"</span><span class="op">)</span></code></pre></div>
<p>The report can be found here: <a href="https://compepigen.github.io/WGBS_best_practice/processed_methrix_reports.html">processed_methrix_reports.html</a></p>
<p>There are additional possibilities to visualize the study. We can look at the density of coverage both sample- and group-wise:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_coverage.html">plot_coverage</a></span><span class="op">(</span><span class="va">meth</span>, type <span class="op">=</span> <span class="st">"dens"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_coverage.html">plot_coverage</a></span><span class="op">(</span><span class="va">meth</span>, type <span class="op">=</span> <span class="st">"dens"</span>, pheno <span class="op">=</span> <span class="st">"Day"</span>, perGroup <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>We can visualize te beta value distribution as a violin plot or density plot. These plots (as well as the coverage plot) use the most variable 25000 CpG sites to ensure fast processing. <code>plot_density</code> and <code>plot_violin</code> has the option to use data only from a restricted region.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_density.html">plot_density</a></span><span class="op">(</span><span class="va">meth</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_density.html">plot_density</a></span><span class="op">(</span><span class="va">meth</span>, ranges <span class="op">=</span> <span class="fu">GRanges</span><span class="op">(</span><span class="st">"chr1:1-10000000"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="dimensionality-reduction" class="section level2">
<h2 class="hasAnchor">
<a href="#dimensionality-reduction" class="anchor"></a>Dimensionality reduction</h2>
<p><code>scMethrix</code> offers principal component analysis (PCA), uniform manifold approximation and projection (UMAP), and t-distributed stochastic neighbor embedding (tSNE) to conduct and visualize dimensionality reduction. The resulting reduction is stored in the scMethrix object, and can be accessed via reducedDim(). First, the model is calculated, then a number of sites are selected for plotting, either random or based on variance (<code>var</code> option). This ensures that the calculations remain feasible.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meth</span> <span class="op">&lt;-</span> <span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="../reference/dim_red_scMethrix.html">dim_red_scMethrix</a></span><span class="op">(</span><span class="va">meth</span>,type<span class="op">=</span><span class="st">"PCA"</span>,top_var <span class="op">=</span> <span class="fl">10000</span>, n_components <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_dim_red.html">plot_dim_red</a></span><span class="op">(</span><span class="va">meth</span>,type<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meth</span> <span class="op">&lt;-</span> <span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="../reference/dim_red_scMethrix.html">dim_red_scMethrix</a></span><span class="op">(</span><span class="va">meth</span>,type<span class="op">=</span><span class="st">"UMAP"</span>,top_var <span class="op">=</span> <span class="fl">10000</span>, n_neighbors <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_dim_red.html">plot_dim_red</a></span><span class="op">(</span><span class="va">meth</span>,type<span class="op">=</span><span class="st">"UMAP"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meth</span> <span class="op">&lt;-</span> <span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="../reference/dim_red_scMethrix.html">dim_red_scMethrix</a></span><span class="op">(</span><span class="va">meth</span>,type<span class="op">=</span><span class="st">"tSNE"</span>,top_var <span class="op">=</span> <span class="fl">10000</span>, n_components <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_dim_red.html">plot_dim_red</a></span><span class="op">(</span><span class="va">meth</span>,type<span class="op">=</span><span class="st">"tSNE"</span><span class="op">)</span></code></pre></div>
<p>At visualization, we can provide the <code>methrix</code> object to allow color or shape annotation of groups or samples.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/methrix/man/plot_pca.html">plot_pca</a></span><span class="op">(</span><span class="va">meth</span>, col_anno <span class="op">=</span> <span class="st">"Day"</span>, shape_anno <span class="op">=</span> <span class="st">"Replicate"</span><span class="op">)</span></code></pre></div>
<p><code>scMethrix</code> offers the possibility of region based filtering. With this option, selected regions, e.g. promoters can be visualized. We will use the <code>AnnotationHub</code> package to assess basic annotation categories.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"AnnotationHub"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"AnnotationHub"</span>, update <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"AnnotationHub"</span><span class="op">)</span>
<span class="va">ah</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html">AnnotationHub</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">qhs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html">query</a></span><span class="op">(</span><span class="va">ah</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"RefSeq"</span>, <span class="st">"Mus musculus"</span>, <span class="st">"mm10"</span><span class="op">)</span><span class="op">)</span>
<span class="va">genes</span> <span class="op">=</span> <span class="va">qhs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="va">promotors</span> <span class="op">=</span> <span class="fu">promoters</span><span class="op">(</span><span class="va">genes</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meth.subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/methrix/man/subset_methrix.html">subset_methrix</a></span><span class="op">(</span><span class="va">meth</span>,regions <span class="op">=</span> <span class="va">promotors</span><span class="op">)</span>
<span class="va">meth</span> <span class="op">&lt;-</span> <span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="../reference/dim_red_scMethrix.html">dim_red_scMethrix</a></span><span class="op">(</span><span class="va">meth.subset</span>,type<span class="op">=</span><span class="st">"PCA"</span>,top_var <span class="op">=</span> <span class="fl">10000</span>, n_components <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_dim_red.html">plot_dim_red</a></span><span class="op">(</span><span class="va">meth</span>,type<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Lindsay Andrew.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
