<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Versatile BedGraph reader. — read_beds • scMethrix</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Versatile BedGraph reader. — read_beds" />
<meta property="og:description" content="Versatile BedGraph reader." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMethrix</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.0001</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/load_data.html">01: Reading in the bedgraph data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/CompEpigen/scMethrix/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Versatile BedGraph reader.</h1>
    <small class="dont-index">Source: <a href='https://github.com/CompEpigen/scMethrix/blob/master/R/read_beds.R'><code>R/read_beds.R</code></a></small>
    <div class="hidden name"><code>read_beds.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Versatile BedGraph reader.</p>
    </div>

    <pre class="usage"><span class='fu'>read_beds</span><span class='op'>(</span>
  files <span class='op'>=</span> <span class='cn'>NULL</span>,
  ref_cpgs <span class='op'>=</span> <span class='cn'>NULL</span>,
  colData <span class='op'>=</span> <span class='cn'>NULL</span>,
  genome_name <span class='op'>=</span> <span class='st'>"hg19"</span>,
  batch_size <span class='op'>=</span> <span class='fl'>20</span>,
  n_threads <span class='op'>=</span> <span class='fl'>0</span>,
  h5 <span class='op'>=</span> <span class='cn'>FALSE</span>,
  h5_dir <span class='op'>=</span> <span class='cn'>NULL</span>,
  h5_temp <span class='op'>=</span> <span class='cn'>NULL</span>,
  desc <span class='op'>=</span> <span class='cn'>NULL</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>,
  zero_based <span class='op'>=</span> <span class='cn'>FALSE</span>,
  reads <span class='op'>=</span> <span class='cn'>NULL</span>,
  replace <span class='op'>=</span> <span class='cn'>FALSE</span>,
  pipeline <span class='op'>=</span> <span class='cn'>NULL</span>,
  stranded <span class='op'>=</span> <span class='cn'>FALSE</span>,
  chr_idx <span class='op'>=</span> <span class='cn'>NULL</span>,
  start_idx <span class='op'>=</span> <span class='cn'>NULL</span>,
  end_idx <span class='op'>=</span> <span class='cn'>NULL</span>,
  beta_idx <span class='op'>=</span> <span class='cn'>NULL</span>,
  M_idx <span class='op'>=</span> <span class='cn'>NULL</span>,
  U_idx <span class='op'>=</span> <span class='cn'>NULL</span>,
  strand_idx <span class='op'>=</span> <span class='cn'>NULL</span>,
  cov_idx <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>files</th>
      <td><p>list of strings; file.paths of BED files</p></td>
    </tr>
    <tr>
      <th>ref_cpgs</th>
      <td><p>data.table; list of CpG sites in the tab-delimited format of chr-start-end. Must be zero-based genome.</p></td>
    </tr>
    <tr>
      <th>colData</th>
      <td><p>list of strings; Sample names. Will be derived from filenames if not provided</p></td>
    </tr>
    <tr>
      <th>genome_name</th>
      <td><p>string; Name of genome. Default hg19</p></td>
    </tr>
    <tr>
      <th>batch_size</th>
      <td><p>integer; Max number of files to hold in memory at once. Default 20</p></td>
    </tr>
    <tr>
      <th>n_threads</th>
      <td><p>integer; number of threads to use. Default 1.
Be-careful - there is a linear increase in memory usage with number of threads. This option is does not work with Windows OS.</p></td>
    </tr>
    <tr>
      <th>h5</th>
      <td><p>boolean; Should the coverage and methylation matrices be stored as <code>HDF5Array</code></p></td>
    </tr>
    <tr>
      <th>h5_dir</th>
      <td><p>string; directory to store H5 based object</p></td>
    </tr>
    <tr>
      <th>h5_temp</th>
      <td><p>string; temporary directory to store hdf5</p></td>
    </tr>
    <tr>
      <th>desc</th>
      <td><p>string; Description of the experiment</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>boolean; flag to output messages or not.</p></td>
    </tr>
    <tr>
      <th>zero_based</th>
      <td><p>boolean; flag for whether the input data is zero-based or not</p></td>
    </tr>
    <tr>
      <th>reads</th>
      <td><p>data.table; Manual input of reads. Typically used for testing.</p></td>
    </tr>
    <tr>
      <th>replace</th>
      <td><p>boolean; flag for whether to delete the contents of h5_dir before saving</p></td>
    </tr>
    <tr>
      <th>pipeline</th>
      <td><p>string; Default NULL. Currently supports "Bismark_cov", "MethylDackel", "MethylcTools", "BisSNP", "BSseeker2_CGmap"
If not known use idx arguments for manual column assignments.</p></td>
    </tr>
    <tr>
      <th>stranded</th>
      <td><p>boolean; Whether in input data is stranded. Default FALSE</p></td>
    </tr>
    <tr>
      <th>chr_idx</th>
      <td><p>integer; column index for chromosome in bedgraph files</p></td>
    </tr>
    <tr>
      <th>start_idx</th>
      <td><p>integer; column index for start position in bedgraph files</p></td>
    </tr>
    <tr>
      <th>end_idx</th>
      <td><p>integer; column index for end position in bedgraph files</p></td>
    </tr>
    <tr>
      <th>beta_idx</th>
      <td><p>integer; column index for beta values in bedgraph files</p></td>
    </tr>
    <tr>
      <th>M_idx</th>
      <td><p>integer; column index for read counts supporting Methylation in bedgraph files</p></td>
    </tr>
    <tr>
      <th>U_idx</th>
      <td><p>integer; column index for read counts supporting Un-methylation in bedgraph files</p></td>
    </tr>
    <tr>
      <th>strand_idx</th>
      <td><p>integer; column index for strand information in bedgraph files</p></td>
    </tr>
    <tr>
      <th>cov_idx</th>
      <td><p>integer; column index for total-coverage in bedgraph files</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An object of class <code><a href='scMethrix-class.html'>scMethrix</a></code></p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Reads BED files and generates methylation matrices.
Optionally arrays can be serialized as on-disk HDFS5 arrays.</p>
<p>colData should be input as a headered data.table with a column called "Sample" with names matching the input filenames. Any other columns may be added to include relevant data (e.g. cell type, collection date, etc). During input, this is done as a left join on the inputted files, so the input colData may contain rows for samples that are not actually included in the analysis. This data will be updated on any relevant subsets or merges, etc.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='co'>#Do Nothing</span>
<span class='op'>}</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Lindsay Andrew.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


