<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>04: Clustering and metacells â€¢ scMethrix</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="04: Clustering and metacells">
<meta property="og:description" content="scMethrix">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMethrix</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.0001</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_load_data.html">01: Reading in the bedgraph data</a>
    </li>
    <li>
      <a href="../articles/02_qc.html">02: QC and reports</a>
    </li>
    <li>
      <a href="../articles/03_transforms.html">03: Transformations and Imputations</a>
    </li>
    <li>
      <a href="../articles/04_clustering.html">04: Clustering and metacells</a>
    </li>
    <li>
      <a href="../articles/05_hdf5.html">05: Working with large cohorts</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/CompEpigen/scMethrix/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="04_clustering_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>04: Clustering and metacells</h1>
                        <h4 class="author">Andrew Lindsay, Reka Toth</h4>
            
            <h4 class="date">2021-09-10</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/CompEpigen/scMethrix/blob/master/vignettes/04_clustering.Rmd"><code>vignettes/04_clustering.Rmd</code></a></small>
      <div class="hidden name"><code>04_clustering.Rmd</code></div>

    </div>

    
    
<div id="clustering-data" class="section level1">
<h1 class="hasAnchor">
<a href="#clustering-data" class="anchor"></a>Clustering data</h1>
<p>Unlike bulk sequencing, where the measurements are averaged over many cells, single-cell sequencing allows the identification of cell type for individual cells. This is a common step in many single-cell expression workflows. Multiple algorithms and distance metrics are included in the package for maximum flexibility.</p>
<p>Most algorithms operate via a distance metric, an all-by-all matrix where each sample is numerically compared to each other sample. Samples with few methylated CpG sites in common will have a large distance metric, where as two samples with identical methylation will have a distance metric of zero.</p>
<p>Available distance metrics include <code>pearson</code>, <code>spearman</code>, and <code>tau</code> (all via <code>bioDist</code>); and <code>euclidean</code>, <code>maximum</code>, <code>manhattan</code>, <code>canberra</code>, <code>binary</code>, and <code>minkowski</code> (all via <code><a href="https://rdrr.io/r/stats/dist.html">stats::dist</a></code>).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dist</span> <span class="op">&lt;-</span> <span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="../reference/get_distance_matrix.html">get_distance_matrix</a></span><span class="op">(</span><span class="va">meth</span>, assay <span class="op">=</span> <span class="st">"score"</span> , type <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span>
<span class="va">dist</span> <span class="op">&lt;-</span> <span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="../reference/get_distance_matrix.html">get_distance_matrix</a></span><span class="op">(</span><span class="va">meth</span>, assay <span class="op">=</span> <span class="st">"score"</span> , type <span class="op">=</span> <span class="st">"pearson"</span><span class="op">)</span>
<span class="va">dist</span> <span class="op">&lt;-</span> <span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="../reference/get_distance_matrix.html">get_distance_matrix</a></span><span class="op">(</span><span class="va">meth</span>, assay <span class="op">=</span> <span class="st">"score"</span> , type <span class="op">=</span> <span class="st">"canberra"</span><span class="op">)</span></code></pre></div>
<p>Clustering algorithms include <code>hierarchical</code> (via <code><a href="https://rdrr.io/r/stats/hclust.html">stats::hclust</a></code>), <code>partition</code> (via <code><a href="https://rdrr.io/r/stats/kmeans.html">stats::kmeans</a></code>), and <code>model</code>-based (via <code><a href="https://mclust-org.github.io/mclust/reference/Mclust.html">mclust::Mclust</a></code>). Distance metrics are not used for <code>model</code>-based clustering. The identified clusters will be stored in the <code>colData</code> slot of the output <code>scMethrix</code> object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meth.cluster</span> <span class="op">&lt;-</span> <span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="../reference/cluster_scMethrix.html">cluster_scMethrix</a></span><span class="op">(</span><span class="va">meth</span>, assay <span class="op">=</span> <span class="st">"score"</span>, dist <span class="op">=</span> <span class="va">dist</span>, n_clusters <span class="op">=</span> <span class="fl">5</span>, colname <span class="op">=</span> <span class="st">"Heirarchical"</span>, type<span class="op">=</span><span class="st">"hierarchical"</span><span class="op">)</span>
<span class="fu">colData</span><span class="op">(</span><span class="va">meth.cluster</span><span class="op">)</span>

<span class="va">meth.cluster</span> <span class="op">&lt;-</span> <span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="../reference/cluster_scMethrix.html">cluster_scMethrix</a></span><span class="op">(</span><span class="va">meth</span>, assay <span class="op">=</span> <span class="st">"score"</span>, dist <span class="op">=</span> <span class="va">dist</span>, colname <span class="op">=</span> <span class="st">"Model-based"</span>, type<span class="op">=</span><span class="st">"model"</span><span class="op">)</span>
<span class="fu">colData</span><span class="op">(</span><span class="va">meth.cluster</span><span class="op">)</span></code></pre></div>
</div>
<div id="visualization" class="section level1">
<h1 class="hasAnchor">
<a href="#visualization" class="anchor"></a>Visualization</h1>
<p>For visualizing the clustering, the generated colData can be used for annotation in dimensionality reduction plots.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meth</span> <span class="op">&lt;-</span> <span class="fu">scMethrix</span><span class="fu">::</span><span class="fu"><a href="../reference/dim_red_scMethrix.html">dim_red_scMethrix</a></span><span class="op">(</span><span class="va">meth</span>,type<span class="op">=</span><span class="st">"PCA"</span>,top_var <span class="op">=</span> <span class="fl">10000</span>, n_components <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_dim_red.html">plot_dim_red</a></span><span class="op">(</span><span class="va">meth</span>,type<span class="op">=</span><span class="st">"PCA"</span>,shape_anno <span class="op">=</span> <span class="st">"Cluster"</span>,color_anno<span class="op">=</span><span class="st">"Cluster"</span><span class="op">)</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Lindsay Andrew.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
