<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>01: Reading in the bedgraph data • scMethrix</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="01: Reading in the bedgraph data">
<meta property="og:description" content="scMethrix">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMethrix</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0.0002</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/x01_load_data.html">01: Reading in the bedgraph data</a>
    </li>
    <li>
      <a href="../articles/x02_qc.html">02: QC and reports</a>
    </li>
    <li>
      <a href="../articles/x03_transforms.html">03: Transformations and Imputations</a>
    </li>
    <li>
      <a href="../articles/x04_clustering.html">04: Clustering and metacells</a>
    </li>
    <li>
      <a href="../articles/x05_hdf5.html">05: Working with large cohorts</a>
    </li>
    <li>
      <a href="../articles/x06_exporting.html">06: Exporting Downstream</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/CompEpigen/scMethrix/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="x01_load_data_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>01: Reading in the bedgraph data</h1>
                        <h4 class="author">Andrew Lindsay, Reka Toth</h4>
            
            <h4 class="date">2021-11-02</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/CompEpigen/scMethrix/blob/master/vignettes/x01_load_data.Rmd"><code>vignettes/x01_load_data.Rmd</code></a></small>
      <div class="hidden name"><code>x01_load_data.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Load library</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/CompEpigen/scMethrix">scMethrix</a></span><span class="op">)</span></code></pre></div>
<div id="download-the-data" class="section level1">
<h1 class="hasAnchor">
<a href="#download-the-data" class="anchor"></a>Download the data</h1>
<p>The example data analysis presented here uses the following publicly available whole genome bisulfite sequencing data from GEO: <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE97179">GSE97179</a></p>
<p>Reference: Luo C, Keown CL, Kurihara L, Zhou J et al. Single-cell methylomes identify neuronal subtypes and regulatory elements in mammalian cortex. Science 2017 Aug 11;357(6351):600-604. <a href="https://www.ncbi.nlm.nih.gov/pubmed/28798132">PMID: 28798132</a></p>
<p>The data can be directly downloaded from GEO, using the <code>GEOquery</code> package. The files will be downloaded into the working directory.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"GEOquery"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"GEOquery"</span><span class="op">)</span>
<span class="op">}</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GEOquery</span><span class="op">)</span>  </code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data_dir</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>,<span class="st">"/GSE131723/"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">data_dir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GSM3814736"</span>,<span class="st">"GSM3814737"</span>,<span class="st">"GSM3814738"</span>,<span class="st">"GSM4830506"</span>,<span class="st">"GSM4830507"</span>,<span class="st">"GSM4830508"</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">gsm</span><span class="op">)</span><span class="op">{</span>
  <span class="fu">GEOquery</span><span class="fu">::</span><span class="fu">getGEOSuppFiles</span><span class="op">(</span>GEO <span class="op">=</span> <span class="va">gsm</span>, baseDir <span class="op">=</span> <span class="va">data_dir</span>, makeDirectory <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
  <span class="co"># bed_dir &lt;- paste0(data_dir,gsm)</span>
  <span class="co"># bed_file &lt;- list.files(bed_dir,full.names=T)</span>
  <span class="co"># R.utils::gunzip(bed_file,destname=paste0(data_dir,gsm,".bed"))</span>
  <span class="co"># unlink(bed_dir,recursive=TRUE,force=TRUE)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bed_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span>
  path <span class="op">=</span> <span class="va">data_dir</span>,
  pattern <span class="op">=</span> <span class="st">"*bedgraph\\.gz$"</span>,
  <span class="co">#pattern = "*.bed$",</span>
  full.names <span class="op">=</span> <span class="cn">TRUE</span>, 
  recursive <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">bed_files</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="processing" class="section level1">
<h1 class="hasAnchor">
<a href="#processing" class="anchor"></a>Processing</h1>
<div id="cpg-annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#cpg-annotation" class="anchor"></a>CpG annotation</h2>
<p>As a first step, we need a list of CpG sites in the respective genome. The CpG sites are listed using the respective <a href="https://bioconductor.org/packages/release/bioc/html/BSgenome.html">BSgenome</a> annotation package. The <code>read_beds</code> function is also able to extract CpG sites on it’s own, however, it might be beneficial to do it separately.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Genome of your preference to work with</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BiocManager</span><span class="op">)</span>
<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span><span class="op">)</span>
<span class="op">}</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg19</span><span class="op">)</span> </code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mm19_cpgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html">suppressWarnings</a></span><span class="op">(</span><span class="fu">extract_CPGs</span><span class="op">(</span>ref_genome <span class="op">=</span> <span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="sample-annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-annotation" class="anchor"></a>Sample annotation</h2>
<p>An annotation table is also necessary to perform some analyses. The data will be added to the <code>scMethrix</code> object, as a <code>colData</code> slot. Each row of the table represents a single sample. Sample names are taken from the input file names, and stored as row.names. Subsequent information about each sample (e.g., cell type, alternate IDs, sample source, etc) are stored as columns within this <code>data.frame</code>. To access this data, see <a href="#saving-experiment-metadata">Saving experiment metadata</a>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sample_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
  row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span>
    pattern <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"(GSM[[:digit:]]*)_epi-gSCAR_(Kasumi-1|OCI-AML3)_SC[[:digit:]]_meth.bedgraph.gz"</span><span class="op">)</span>,
    replacement <span class="op">=</span> <span class="st">"\\1"</span>,
    x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">bed_files</span><span class="op">)</span>
    <span class="op">)</span>,
  Group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span>
    pattern <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"(GSM[[:digit:]]*)_epi-gSCAR_(Kasumi-1|OCI-AML3)_SC[[:digit:]]_meth.bedgraph.gz"</span><span class="op">)</span>,
    replacement <span class="op">=</span> <span class="st">"\\2"</span>,
    x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">bed_files</span><span class="op">)</span>
    <span class="op">)</span>,
  stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span>
<span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">sample_anno</span><span class="op">)</span></code></pre></div>
</div>
<div id="reading-bedgraph-files" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-bedgraph-files" class="anchor"></a>Reading bedGraph files</h2>
<p><code>read_beds</code> function is a versatile bedgraph reader intended to import bedgraph files generated virtually by any sort of methylation calling program. It requires user to provide indices for chromosome names, start position and other required fields. There are also presets available to import <code>bedgraphs</code> from most common programs such as <code>Bismark</code> (.cov format), <code>MethylDackel</code>, and <code>MethylcTools</code>. In this case, there is no need to define e.g. <code>chr_idx</code> <code>start_idx</code> arguments, the function will automatically assign them. To know the exact parameters, it might worth to take a look at one of the files, to see the column order.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">fread</span><span class="op">(</span><span class="va">bed_files</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chr"</span>,<span class="st">"start"</span>,<span class="st">"end"</span>,<span class="st">"beta"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></code></pre></div>
<p>The <code>read_beds</code> function adds CpGs missing from the reference set, and creates a methylation/coverage matrices. Once the process is complete - it returns an object of class <code>scMethrix</code> which in turn inherits <code>SingleCellExperiment</code> class. <code>scMethrix</code> object contains ‘methylation’ and (optionally) ‘coverage’ matrices (either in-memory or as on-disk HDF5 arrays) along with pheno-data and other basic info. This object can be passed to all downstream functions for various analysis. For further details on the data structure, see the <code>SingleCellExperiment</code> package: <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html">SingleCellExperiment</a></p>
<ul>
<li>For computers with large memory, increase speed using <code>batch_size</code> parameter. This will open multiple files at once, but keep in mind that increasing the batch size will increase the memory needs as well.</li>
</ul>
<p>A more detailed description of the <code>read_beds</code> function is here: <a href="https://compepigen.github.io/WGBS_best_practice/read_bedgraphs.html">read_beds</a>)</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meth</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_beds.html">read_beds</a></span><span class="op">(</span>
  files <span class="op">=</span> <span class="va">bed_files</span>,
  ref_cpgs <span class="op">=</span> <span class="va">mm19_cpgs</span>,
  chr_idx <span class="op">=</span> <span class="fl">1</span>,
  start_idx <span class="op">=</span> <span class="fl">2</span>,
  strand_idx <span class="op">=</span> <span class="fl">3</span>,
  cov_idx <span class="op">=</span> <span class="fl">4</span>,
  M_idx <span class="op">=</span> <span class="fl">5</span>,
  stranded <span class="op">=</span> <span class="cn">FALSE</span>,
  zero_based <span class="op">=</span> <span class="cn">TRUE</span>, 
  <span class="co">#collapse_strands = FALSE, </span>
  colData <span class="op">=</span> <span class="va">sample_anno</span>
<span class="op">)</span></code></pre></div>
<p>Certain operations will cause metadata to be destroyed. For range metadata, <code><a href="../reference/bin_scMethrix.html">bin_scMethrix()</a></code> will discard row metadata, as will <code>collapse_samples</code> for column (sample) metadata. It is advised to extract this data from <code>mcols()</code> and <code>colData()</code>, respectively, and store it within <code>metadata()</code> instead.</p>
<p><strong>Potential pitfalls: </strong> Too many missing CpGs, if not expected, might indicate that something went wrong.</p>
<ul>
<li><p>An obvious error is using the wrong reference genome.</p></li>
<li><p>A common mistake is falsely indicating strandedness and zero or one based positioning. The combination of the two might lead to only using C-s from one strand (due to a 1 bp shift), resulting lower coverage for the sites. A nice description of the zero- and one-based coordination system can be found <a href="http://genome.ucsc.edu/blog/the-ucsc-genome-browser-coordinate-counting-systems/">here</a></p></li>
</ul>
</div>
<div id="extracting-assay-matricies" class="section level2">
<h2 class="hasAnchor">
<a href="#extracting-assay-matricies" class="anchor"></a>Extracting assay matricies</h2>
<p>After the experiment is created, an assay matrix can be generated. Genomic loci can also be added (<code>add_loci</code>), be formatted as a <code>GenomicRanges</code> object if desired, and ordered by variance (<code>order_by_sd</code>). Additionally, the methylation and coverage matrices can be directly obtained with the functions <code>score()</code> and <code>counts()</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mtx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_matrix.html">get_matrix</a></span><span class="op">(</span><span class="va">meth</span>, assay<span class="op">=</span><span class="st">"score"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">mtx</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Add genomic loci, put in a GRanges object, and order by SD</span>
<span class="va">mtx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_matrix.html">get_matrix</a></span><span class="op">(</span><span class="va">meth</span>, assay<span class="op">=</span><span class="st">"score"</span>, add_loci <span class="op">=</span> <span class="cn">TRUE</span>, in_granges<span class="op">=</span><span class="cn">TRUE</span>, order_by_sd <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">mtx</span><span class="op">)</span></code></pre></div>
</div>
<div id="saving-experiment-metadata" class="section level2">
<h2 class="hasAnchor">
<a href="#saving-experiment-metadata" class="anchor"></a>Saving experiment metadata</h2>
<p>scMethrix can store experiment metadata. Depending on the type of data, it is stored in different places:</p>
<ol style="list-style-type: decimal">
<li>Experiment-wide metadata: General information is stored in metadata() as a named <code>list</code>. This can be information like genome accession, GEO IDs, authors, or any other information that needs just a single variable.</li>
</ol>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">metadata</span><span class="op">(</span><span class="va">scm</span><span class="op">)</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Range metadata: Information about ranges is stored as a <code>data.frame</code> in mcols(). This is stored within the rowRanges() sub-object inside</li>
</ol>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">mcols</span><span class="op">(</span><span class="va">scm</span><span class="op">)</span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Sample metadata: Information about samples is stored as a <code>data.frame</code> in colData(). This can include alternate samples IDs, run IDs, tissue types, or anything associated with individual samples.</li>
</ol>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">scm</span><span class="op">)</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Lindsay Andrew.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
